lib/factory_bot/definition.rb

```diff
@@ -114,14 +114,28 @@ def trait_by_name(name)
       trait_for(name) || Internal.trait_by_name(name)
     end
 
+    # This is a hot path when using traits so:
+    # - Memoize the defined_traits by name so we can lookup in a hash instead of array
+    # - Store both string/symbol names to avoid needing to cast on lookup
+    # - Use plain Hash since HashWithIndifferentAccess incurs significant overhead
     def trait_for(name)
-      defined_traits.detect { |trait| trait.name == name.to_s }
+      @defined_traits_by_name ||= defined_traits.each_with_object({}) do |trait, memo|
+        # This method was previously using `detect` which means if there were
+        # multiple traits the first was returned. It doesn't seem like you should be able
+        # to register duplicate traits but without this logic a test was breaking:
```

Comment:
This reminds me of https://github.com/thoughtbot/factory_bot/issues/968. I think there was talk at once point of using inheritance for traits instead of copying them all over to the child factories.